ui/app/selectors/transactions.js

```diff
@@ -251,11 +253,16 @@ export const nonceSortedTransactionsSelector = createSelector(
           nonceProps.hasCancelled = true
         }
       } else {
+        let primaryTransaction = transaction
+        if (transaction.status === FAILED_STATUS && Boolean(transaction?.err?.message)) {
```

Comment:
The only place where `transaction.err` is set is in the `tx-state-manager`'s `setTxStatusFailed` method. 

this method is only called inside of the main transaction controller. It's set when a transaction fails to transition from an unapproved transaction to an approved one (the user has confirmed the transaction, which queues it up to be sent to the network).

There are two more cases where this method is called, though, both of which are in the `_onBootCleanUp` method of the transaction controller. This method operates on two distinct categories of transactions: approved transactions and unapproved ones that have the 'loadingDefaults' flag. In the latter case, the intent is to add default gas estimations to the unapproved transactions so that they can be sent to the user for approval. In the former case, these transactions are automatically moved to the failed state. It appears we might not be able to effectively transition approved transactions that do not immediately get moved to the submitted status if the background process somehow terminates/restarts. 

Those cases don't appear to me to be things that we'd want to reassign `primaryTransaction` for, either. So I think this change is safe. I'd like @Gudahtt, @rekmarks, or @whymarrh to double-check me on this though.

Reply:
Either way I think this is the wrong property to check. See here: https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/metamask-controller.js#L283

An on-chain failure might also be represented by a "confirmed" transaction with an error status on the transaction receipt. There would be no `err` property in that case. In the MetaMask Controller, we check for a status of either "failed" or "confirmed", and we look for this zero status on the receipt. [A status of zero denotes an on-chain failure.](https://eth.wiki/json-rpc/API#eth_gettransactionreceipt)

Reply:
I actually don't fully understand the case where a transaction has a status of "failed" _and_ a transaction receipt. Maybe that's not possible? It is probably safe to assume that _if_ it has a transaction receipt with a status of zero, then it must be an on-chain failure though.

Reply:
Would that confirmed transaction with an on-chain failure be a 'failed' or 'confirmed' status in our state?

Reply:
Also, just to make sure I understand fully, are you suggesting that we look at `txReceipt` to determine failure?

Reply:
This status check is here as well: https://github.com/MetaMask/metamask-extension/blob/develop/ui/app/helpers/utils/transactions.util.js#L172
This is how the UI differentiates between confirmed successful transactions and on-chain failures.

Reply:
In the background, an on-chain failure _should_ have a status of "confirmed" with a `txReceipt` status of `0x0`. But I don't understand all of the potential failure cases, so I can't say what might happen if an error is thrown somewhere. It's possible we have on-chain failures on transactions with a "failed" status as well.

Reply:
>Also, just to make sure I understand fully, are you suggesting that we look at `txReceipt` to determine failure?

`txMeta.txReceipt.status` is the sole way we should be checking for on-chain failures. I think you're correct about `txMeta.err` being the best way to check for _other_ error states.

Reply:
Oops. When I said "Either way I think this is the wrong property to check", I was confused and thinking we were trying to detect on-chain failures. Really we're trying to detect failures that _aren't_ on-chain. Sorry :sweat_smile:

I still think we should check the `txReceipt.status` to filter out on-chain failures though, just on the off chance that they _somehow_ got a status of "failed" instead of "confirmed".

Reply:
Seems like we should be checking for on chain failures in the controller instead of having many different branches of failure that have to be checked in the UI.

Reply:
With that said we'd need to check the primaryTransaction `status === FAILED && transaction.err` or `status === CONFIRMED && transaction.txReceipt.status === '0x0'` if that's true, override `primaryTransaction` with current pending/confirmed transaction (as long as it also is not an on-chain failure)

Reply:
@darkwing and @Gudahtt lets jump on a call on Monday to talk through this? It'd be helpful for me at least as I continue working on activity list refactor.